src/util/metadata/header_metadata.rs

```diff
@@ -29,668 +27,1014 @@ pub struct HeaderMetadataSpec {
-                .fetch_add(val as u8, order)
-                .into()
+    /// This function provides a default implementation for the `fetch_add` method from the `ObjectModel` trait.
+    #[inline(always)]
+    pub fn fetch_add<T: MetadataValue>(
+        &self,
+        object: ObjectReference,
+        val: T,
+        order: Ordering,
+    ) -> T {
+        #[cfg(debug_assertions)]
+        self.assert_spec::<T>();
+        if self.num_of_bits < 8 {
+            FromPrimitive::from_u8(self.fetch_ops_on_bits(object, order, order, |x: u8| {
+                x.wrapping_add(val.to_u8().unwrap())
+            }))
+            .unwrap()
+        } else {
+            unsafe { T::fetch_add(self.meta_addr(object), val, order) }
         }
-    } else if metadata_spec.num_of_bits == 16 {
-        debug_assert!(
-            metadata_spec.bit_offset.trailing_zeros() as usize >= LOG_BITS_IN_U16,
-            "Metadata 16-bits: ({:?}) bit_offset must be 2-bytes aligned!",
-            metadata_spec
-        );
-        let u16_offset = metadata_spec.bit_offset >> LOG_BITS_IN_BYTE;
+    }
 
-        unsafe {
-            (*(object.to_address() + u16_offset).to_ptr::<AtomicU16>())
-                .fetch_add(val as u16, order)
-                .into()
+    /// This function provides a default implementation for the `fetch_sub` method from the `ObjectModel` trait.
+    #[inline(always)]
+    pub fn fetch_sub<T: MetadataValue>(
+        &self,
+        object: ObjectReference,
+        val: T,
+        order: Ordering,
+    ) -> T {
+        #[cfg(debug_assertions)]
+        self.assert_spec::<T>();
+        if self.num_of_bits < 8 {
+            FromPrimitive::from_u8(self.fetch_ops_on_bits(object, order, order, |x: u8| {
+                x.wrapping_sub(val.to_u8().unwrap())
+            }))
+            .unwrap()
+        } else {
+            unsafe { T::fetch_sub(self.meta_addr(object), val, order) }
         }
-    } else if metadata_spec.num_of_bits == 32 {
-        debug_assert!(
-            metadata_spec.bit_offset.trailing_zeros() as usize >= LOG_BITS_IN_U32,
-            "Metadata 32-bits: ({:?}) bit_offset must be 4-bytes aligned!",
-            metadata_spec
-        );
-        let u32_offset = metadata_spec.bit_offset >> LOG_BITS_IN_BYTE;
+    }
 
-        unsafe {
-            (*(object.to_address() + u32_offset).to_ptr::<AtomicU32>()).fetch_add(val as u32, order)
-                as usize
+    /// This function provides a default implementation for the `fetch_and` method from the `ObjectModel` trait.
+    #[inline(always)]
+    pub fn fetch_and<T: MetadataValue>(
+        &self,
+        object: ObjectReference,
+        val: T,
+        order: Ordering,
+    ) -> T {
+        #[cfg(debug_assertions)]
+        self.assert_spec::<T>();
+        if self.num_of_bits < 8 {
+            FromPrimitive::from_u8(
+                self.fetch_ops_on_bits(object, order, order, |x: u8| x & val.to_u8().unwrap()),
```

Comment:
Similar to side metadata, fetch_and doesn't need a loop.

Reply:
Done.